### PR TITLE
cuGraph-PyG Loader Improvements

### LINKED ISSUE #3807: cuGraph-PyG Should Test Training in CI
Currently, we only test compatibility with models as part of benchmarking.  In order to catch issues before they appear there, we should have a basic training test that passes in output samples from cuGraph-PyG into a model.  This would be mostly a smoke test; the other tests should verify that the output of cuGraph-PyG is correct, regardless.

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Consolidates various speed improvements tested while running performance benchmarks.  Avoids copying batch data, removes redundant data loading code, simplifies and improves de-offsetting, even though that is now being bypassed entirely for homogeneous graphs.  Removes extra host to device copy.  Properly flips the src/dst columns in the returned `HeteroData` minibatch objects, avoid exposing this to the end user.

I've confirmed this cuts the MFG time by a factor of 4.

Closes #3807

### CODE REVIEW

**Review by VibhuJawa — COMMENTED** [review]
The changes look good to me, just have asked for a  bunch of clarifications.

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def _sampler_output_from_sampling_results(
 
     hops = torch.arange(sampling_results.hop_id.max() + 1, device="cuda")
     hops = torch.searchsorted(
-        torch.as_tensor(sampling_results.hop_id.values, device="cuda"), hops
+        torch.as_tensor(sampling_results.hop_id, device="cuda"), hops
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:112]: This does use `cuda_array_interface`.  https://github.com/pytorch/pytorch/blob/3a3cf0e09d475df9237c95ebd14debf650e0c038/torch/csrc/utils/tensor_new.cpp#L313

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py [hunk 2]
```diff
@@ -200,35 +200,61 @@ def _sampler_output_from_sampling_results(
...
+                    )
+                    .unique()
+                    .shape[0]
+                )
             else:
-                num_unique_nodes = _count_unique_nodes(
-                    sampling_results_hop, graph_store, node_type, "dst"
+                unique_nodes_hop = _get_unique_nodes(
+                    sampling_results_to_hop, graph_store, node_type, "dst"
                 )
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:224]: I forgot about the ordering constraints in PyG, so I think your solution for DGL will actually work, with a few tweaks.  I'll update my code and get some numbers to see if it is faster.

File: python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py [hunk 3]
```diff
@@ -265,43 +270,48 @@ def __next__(self):
         f = self.__data["batch_id"] == self.__next_batch
         if self.__renumber_map is not None:
             i = self.__next_batch - self.__start_inclusive
-            ix = self.__renumber_map.iloc[[i, i + 1]]
-            ix_start, ix_end = ix.iloc[0], ix.iloc[1]
-            current_renumber_map = self.__renumber_map.iloc[ix_start:ix_end]
+            ix_start, ix_end = self.__renumber_map[[i, i + 1]].tolist()
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py:273]: It's a device tensor.

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py [hunk 4]
```diff
@@ -429,20 +452,48 @@ def __construct_graph(
...
+            # Have to check for empty partitions and handle them appropriately
+            df = (
+                df.map_partitions(
+                    lambda f: cudf.DataFrame.from_pandas(f)
+                    if len(f) > 0
+                    else get_empty_df(),
+                    meta=get_empty_df(),
+                )
+                .reset_index(drop=True)
+                .persist()
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:493]: Is there a better way to do this?  This function already assumes that the data fits on local host memory.  It's not designed to handle the 100B edge scale, just to be a simpler way for users to handle SNMG graphs.

[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:493]: I normally just persist the frame on host memory  and then cache the cugraph.Graph object (which i am sure you do somewhere too).

Reduces 1x overhead while creating cugraph.Graph which is often the bottleneck enabling users to run workflows. 

```suggestion
                df = df.persist()
                df = (
                df.map_partitions(
                    lambda f: cudf.DataFrame.from_pandas(f)
                    if len(f) > 0
                    else get_empty_df(),
                    meta=get_empty_df(),
                )
                .reset_index(drop=True)
```

File: python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py:273]: ok

**Comment by alexbarghi-nv** [comment]
Based on feedback from @VibhuJawa , this PR is going back to draft while I work on improvements

File: python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py]: I switched to storing the offsets separately instead of copying to host.  See code comments.

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py]: I've implemented the solution, but I had to split heterogeneous processing into its own function to simplify things.

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py [hunk 5]
```diff
@@ -256,6 +258,12 @@ def __init__(
         multi_gpu: bool (Optional, default = False)
             Whether the store should be backed by a multi-GPU graph.
             Requires dask to have been set up.
+
+        order: bool (Optional ["CSR", "CSC"], default = CSC)
```
Comment (BradReesWork) [inline python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:262]: order does not look like a bool

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py [hunk 6]
```diff
@@ -78,9 +78,101 @@ def _count_unique_nodes(
...
+def _sampler_output_from_sampling_results_homogeneous(
+    sampling_results: cudf.DataFrame,
+    renumber_map: torch.Tensor,
+    graph_store: CuGraphStore,
+    data_index: Dict[Tuple[int, int], int],
+    batch_id: int,
+    metadata: Sequence = None,
+) -> HeteroSamplerOutput:
+    """
+    Parameters
```
Comment (BradReesWork) [inline python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:95]: the list of parameters does not match what is in the function signature

**Review by BradReesWork — CHANGES_REQUESTED** [review]

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py (hunk 5 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py]: Fixed.

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py [hunk 7]
```diff
@@ -788,25 +841,52 @@ def _get_renumbered_edge_groups_from_sample(
...
+            if len(self.__vertex_type_offsets["type"]) == 1:
+                vtype = src_type
+                id_table = noi_index[vtype]
+                id_map = cudf.Series(
+                    cupy.arange(id_table.shape[0], dtype="int32"),
+                    name="new_id",
+                    index=cupy.asarray(id_table),
+                ).sort_index()
+
+                ix_r = torch.searchsorted(
```
Comment (VibhuJawa) [inline python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:853]: Can we add comments here to help understand what is happening in this if/elese case. Not sure how the logic is flowing here.

File: python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py [hunk 8]
```diff
@@ -253,55 +256,97 @@ def __next__(self):
...
             )
         else:
-            out = torch_geometric.loader.utils.filter_custom_store(
-                self.__feature_store,
-                self.__graph_store,
-                sampler_output.node,
-                sampler_output.row,
-                sampler_output.col,
-                sampler_output.edge,
+            sampler_output = _sampler_output_from_sampling_results(
```
Comment (VibhuJawa) [inline python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py:317]: I would strongly suggest renaming this function to `heterogeneous` to improve code readability.  
```suggestion
            sampler_output = _sampler_output_from_sampling_results_heterogeneous(
```

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Code changes mostly look good, requested some changes to improve code clarity and readability.

File: python/cugraph-pyg/cugraph_pyg/tests/test_cugraph_loader.py [hunk 9]
```diff
@@ -159,33 +184,45 @@ def test_cugraph_loader_from_disk_subset():
...
+            == bogus_samples.destinations.dropna().values_host.tolist()
+        )
 
     assert num_samples == 100
 
 
 @pytest.mark.skipif(isinstance(torch, MissingModule), reason="torch not available")
-def test_cugraph_loader_from_disk_subset_renumbered():
+def test_cugraph_loader_e2e_coo():
+    m = [2, 9, 99, 82, 9, 3, 18, 1, 12]
```
Comment (VibhuJawa) [inline python/cugraph-pyg/cugraph_pyg/tests/test_cugraph_loader.py:206]: I really like this test and we should add this for cugraph-dgl too.

Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/tests/test_cugraph_loader.py:206]: Yeah, this test caught several bugs immediately after I added it.

File: python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py (hunk 8 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/loader/cugraph_node_loader.py:317]: done

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py (hunk 7 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py]: https://github.com/rapidsai/cugraph/issues/3836

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Review by rlratzel — COMMENTED** [review]
I scanned it and only found two things that jumped out to me.  Vibhu and Tingyu would be able to provide a more thorough review.

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py [hunk 10]
```diff
@@ -134,112 +231,87 @@ def _sampler_output_from_sampling_results(
...
+    )
+    del nodes_of_interest
 
-        # Get the new edge index (by type as expected for HeteroData)
-        # FIXME handle edge ids/types after the C++ updates
-        row_dict, col_dict = graph_store._get_renumbered_edge_groups_from_sample(
-            sampling_results, noi_index
-        )
+    # Get the new edge index (by type as expected for HeteroData)
+    # FIXME handle edge ids/types after the C++ updates
```
Comment (acostadon) [inline python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:269]: Are c++ updates are in another PR? or scheduled? What does notwork in the meantime?

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py [hunk 11]
```diff
@@ -822,36 +920,18 @@ def _get_renumbered_edge_groups_from_sample(
...
-                    .reset_index()
-                    .rename(columns={"index": "new_id"})
-                    .set_index("src")
-                )
-                src = src_id_map["new_id"].loc[cupy.asarray(sources)]
-                row_dict[pyg_can_edge_type] = torch.as_tensor(src.values, device="cuda")
+                if self.__order == "CSR":
+                    src_type, _, dst_type = pyg_can_edge_type
+                else:  # CSC
+                    dst_type, _, src_type = pyg_can_edge_type
```
Comment (acostadon) [inline python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:928]: you don't need an error if order is a bad value?

**Review by acostadon — APPROVED** [review]
Looks great as much as I understood. Just saw aplace where I wasn't sure what wasn't working due to the fixme

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py [hunk 12]
```diff
@@ -785,29 +839,73 @@ def _get_renumbered_edge_groups_from_sample(
...
-            )
-            src = src_id_map["new_id"].loc[sampling_results.sources]
-            row_dict[t_pyg_type] = torch.as_tensor(src.values, device="cuda")
+            # If there is only 1 node type (homogeneous)
+            # This should only occur if the cuGraph loader was
+            # not used.  This logic is deprecated.
+            if len(self.node_types) == 1:
+                warnings.warn(
+                    "Renumbering after sampling for homogeneous graphs is deprecated.",
+                    DeprecationWarning,
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py]: fixed

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py (hunk 10 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:269]: Future PR that I am currently working on.  Right now everything works, but we don't have all the promised features for version 23.10 that take advantage of the C++ MFG creation and format conversion.

File: python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py (hunk 11 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/cugraph_store.py:928]: added checking to `__init__`

File: python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py [hunk 13]
```diff
@@ -78,12 +78,111 @@ def _count_unique_nodes(
...
+                    max_id_hop - max_id_prev_hop
+                )
+            else:
+                num_nodes_per_hop_dict[node_type][hop + 1] = 0
+        # will default to 0 if the previous hop was 0, since this is a PyG requirement
+
+        num_edges_per_hop_dict[edge_type][hop] = hop_ix_end - hop_ix_start
+
+    if HeteroSamplerOutput is None:
+        raise ImportError("Error importing from pyg")
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/sampler/cugraph_sampler.py:172]: fixed

**Review by BradReesWork — APPROVED** [review]
